### PR TITLE
Fix #3531 Add collapse/pin/hide functionalities to the widgets

### LINKED ISSUE #3531: Widgets collapse/pin/hide functionalities
### Acceitance Criteria
- [x] The user must be able to collapse/expand widgets: one by one or all together, in a tray
- [x] The admin should be able to pin widget: a pinned widget can not be moved, resized or collapsed. A normal user can not pin/unpin widgets.
- [x] when MapStore is configured accordingly in `localConfig.json`, certain users should be able hide widgets. An hidden user is not visible to other users. 
 - [x] All the properties and the visibility of these tools should be configurable by user role. 
 - [x] User should also be able to add images to the text widget content

### PR DESCRIPTION
**Pull request by offtherailz**
## Description
With this changes:
 - widgets can be collapsed in the new "WidgetTray" plugin. 
 - widgets can be pinned (see the related issue for definition of `pin`)
 - widgets can be hidden to certain users (via localConfig.json)
 - hide and pin tools visibility and policies can be configured in localConfig.json. 
 - text widgets can contain images

### Pin
These tools are implemented as icons on the top left corner of a widget
image
Pinned widgets look like this: 
image
The pin tool is not visible to users that are not able to pin/unpin widgets. By default, only admin can pin (configurable). 

### Collapse
Collapsed widget appear in a tray
image
When collapse a widget, the tray expands. You can collapse all widgets at once using the "collapse all" button in the tray

### Hide
The hide functionality is disabled by default. If configured (in localConfig.json), a user that belongs to a certain role can be able to hide a widget to other users. 

### Add image to text widget
When you edit/create a widget, the WYSIWYG editor have the tool to add images. 

image

## Issues
 - Fix #3531
 - Connected to #2655 (only for collapse widget task)

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Feature

**What is the current behavior?** (You can also link to an open issue here)
Nor of features in  #3531 is present

**What is the new behavior?**
Widgets are collapsible

**Does this PR introduce a breaking change?** (check one with "x", remove the other)

 - [x] No

**Other information**:

### CODE REVIEW

File: web/client/actions/widgets.js [hunk 1]
```diff
@@ -248,6 +251,29 @@ const toggleDependencySelector = (active, settings) => setupDependencySelector({
...
+const toggleCollapse = (widget, target = DEFAULT_TARGET) => ({
+    type: TOGGLE_COLLAPSE,
+    widget,
+    target
+});
+
+/**
+ * Collapse/Expand all the widgets
+ */
+const toggleCollapseAll = (target = DEFAULT_TARGET) => ({
```
Comment (MV88) [inline web/client/actions/widgets.js]: test for this

File: web/client/actions/widgets.js [hunk 2]
```diff
@@ -248,6 +251,29 @@ const toggleDependencySelector = (active, settings) => setupDependencySelector({
     active,
     ...settings
 });
+/**
+ * Collapse/Expand the widget
+ * @param {object} widget the widget to collapse
+ */
+const toggleCollapse = (widget, target = DEFAULT_TARGET) => ({
```
Comment (MV88) [inline web/client/actions/widgets.js]: test for this, i know it is an action but you choose

File: web/client/components/misc/enhancers/security/accessRuleParser.jsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
+/*
```
Comment (MV88) [inline web/client/components/misc/enhancers/security/accessRuleParser.jsx]: should we move this in components/security/enhancers?

File: web/client/components/widgets/enhancers/tools/pinnableWidget.js [hunk 4]
```diff
@@ -0,0 +1,27 @@
+/*
+ * Copyright 2018, GeoSolutions Sas.
```
Comment (MV88) [inline web/client/components/widgets/enhancers/tools/pinnableWidget.js]: 2019 :blush:

Comment (offtherailz) [reply web/client/components/widgets/enhancers/tools/pinnableWidget.js]: Some of these files has been created in 2018 :smile:

File: web/client/actions/widgets.js (hunk 1 above)
Comment (offtherailz) [reply web/client/actions/widgets.js]: It's tested with the reducer, anyway, I need to add a test for the action itself

File: web/client/actions/widgets.js (hunk 2 above)
Comment (offtherailz) [reply web/client/actions/widgets.js]: tested with the reducer, anyway, I need to add a test for the action itself

File: web/client/components/misc/enhancers/security/accessRuleParser.jsx (hunk 3 above)
Comment (offtherailz) [reply web/client/components/misc/enhancers/security/accessRuleParser.jsx]: security folder is now dedicated to user, group, and permission editor. This is a different thing, so I'm afraid placing it there may be misleading

**Pull request merged**